File: tools/event_prep/us_twt_irrigation_schedule.py
"""Replace the flood-on / drain-off pair with a repeating irrigation schedule.

SIPNET has no drain event type and no persistent flood state. An irrigation event
adds its amount to the soil water pool once and nothing else
(https://github.com/PecanProject/sipnet/blob/master/src/sipnet/events.c, case
IRRIGATION: soilAmount = amount - evapAmount; fluxes.eventSoilWater += soilAmount).
So a 0 mm event is a no-op that still writes a row to events.out, and a single
150 mm application cannot hold a flood.

The flooded period is instead represented the way the 2017-2023 monitoring record
represents it: repeated applications through the season. Rate and interval come
from those events - 129 flood-method applications, mean 69.7 mm, median gap 8 d,
modal gap 7 d. Drainage becomes the absence of further irrigation, which is what
waterDrainFrac then acts on.

The WTD-derived flood and drain dates are kept as the bounds of each schedule, so
that derivation is preserved rather than discarded.

Usage:  us_twt_irrigation_schedule.py <events.json>   (edits in place)
"""
import json, datetime, copy, sys

IRRIGATION_MM = 70.0   # 2017-2023 mean 69.7 mm
INTERVAL_D = 7         # 2017-2023 modal gap 7 d (median 8 d)
RUN_START = datetime.date(2010, 1, 1)
RUN_END = datetime.date(2016, 12, 31)

path = sys.argv[1] if len(sys.argv) > 1 else sys.exit(__doc__)
doc = json.load(open(path))
ev = doc["events"]
D = datetime.date.fromisoformat

irr = sorted([e for e in ev if e["event_type"] == "irrigation"], key=lambda e: e["date"])
other = [e for e in ev if e["event_type"] != "irrigation"]

# pair each flood with the next drain, in date order; a leading drain is a carry-in
spans, pending = [], None
for e in irr:
    if e["amount_mm"] > 0:
        if pending is not None:
            spans.append((pending, None))
        pending = e
    else:
        spans.append((pending, e))
        pending = None
if pending is not None:
    spans.append((pending, None))

new = []
for flood, drain in spans:
    start = D(flood["date"]) if flood else RUN_START
    end = D(drain["date"]) if drain else RUN_END
    if end <= start:
        print(f"  SKIP non-positive span {start} -> {end}", file=sys.stderr)
        continue
    tmpl = flood or drain
    n, d = 0, start
    while d < end:
        e = copy.deepcopy(tmpl)
        e["date"] = d.isoformat()
        e["amount_mm"] = IRRIGATION_MM
        e["method"] = "flood"
        if n == 0 and flood:
            e["provenance_note"] = (
                flood.get("provenance_note", "") +
                f" Flood held by repeated application of {IRRIGATION_MM:.0f} mm every "
                f"{INTERVAL_D} d, the rate and interval of the 2017-2023 monitoring record, "
                f"because SIPNET has no persistent flood state."
            ).strip()
        elif n == 0:
            e["provenance_note"] = (
                f"Carry-in of the winter flood established before the run window; held at "
                f"{IRRIGATION_MM:.0f} mm every {INTERVAL_D} d until the WTD-derived drain."
            )
        else:
            e["provenance_note"] = (
                f"Maintains the flooded period begun {start.isoformat()}; ends "
                f"{end.isoformat()}, the WTD-derived drain date." if drain else
                f"Maintains the flooded period begun {start.isoformat()} to the end of the run window."
            )
        new.append(e)
        n += 1
        d += datetime.timedelta(days=INTERVAL_D)
    print(f"  {start} -> {end}  ({(end-start).days:3d} d)  {n:2d} events")

doc["events"] = sorted(other + new, key=lambda e: (e["date"], e["event_type"]))
json.dump(doc, open(path, "w"), indent=1)
print(f"\n  irrigation: {len(irr)} -> {len(new)}   total: {len(ev)} -> {len(doc['events'])}")

File: tools/event_prep/us_twt_irrigation_schedule.R
#!/usr/bin/env Rscript
# Replace the flood-on / drain-off pair with a repeating irrigation schedule, since
# SIPNET has no drain event and a 0 mm irrigation is a no-op.

suppressMessages(library(jsonlite))

IRRIGATION_MM <- 70    # 2017-2023 mean 69.7 mm
INTERVAL_D <- 7        # 2017-2023 modal gap 7 d (median 8 d)
RUN_START <- as.Date("2010-01-01")
RUN_END <- as.Date("2016-12-31")

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 1L) stop("usage: us_twt_irrigation_schedule.R <events.json>")
path <- args[1]

doc <- jsonlite::fromJSON(path, simplifyDataFrame = FALSE)
ev <- doc$events
is_irr <- vapply(ev, function(e) identical(e$event_type, "irrigation"), logical(1))
irr <- ev[is_irr]
irr <- irr[order(vapply(irr, function(e) e$date, character(1)))]
other <- ev[!is_irr]

# pair each flood with the next drain, in date order; a leading drain is a carry-in
spans <- list(); pending <- NULL
for (e in irr) {
  if (e$amount_mm > 0) {
    if (!is.null(pending)) spans[[length(spans) + 1L]] <- list(flood = pending, drain = NULL)
    pending <- e
  } else {
    spans[[length(spans) + 1L]] <- list(flood = pending, drain = e)
    pending <- NULL
  }
}
if (!is.null(pending)) spans[[length(spans) + 1L]] <- list(flood = pending, drain = NULL)

new <- list()
for (sp in spans) {
  start <- if (!is.null(sp$flood)) as.Date(sp$flood$date) else RUN_START
  end <- if (!is.null(sp$drain)) as.Date(sp$drain$date) else RUN_END
  if (end <= start) {
    message(sprintf("  SKIP non-positive span %s -> %s", start, end)); next
  }
  tmpl <- if (!is.null(sp$flood)) sp$flood else sp$drain
  n <- 0L; d <- start
  while (d < end) {
    e <- tmpl
    e$date <- format(d, "%Y-%m-%d")
    e$amount_mm <- IRRIGATION_MM
    e$method <- "flood"
    e$provenance_note <- if (n == 0L && !is.null(sp$flood)) {
      trimws(paste0(
        if (is.null(tmpl$provenance_note)) "" else tmpl$provenance_note,
        sprintf(paste("  Flood held by repeated application of %d mm every %d d,",
                      "the rate and interval of the 2017-2023 monitoring record,",
                      "because SIPNET has no persistent flood state."),
                IRRIGATION_MM, INTERVAL_D)))
    } else if (n == 0L) {
      sprintf(paste("Carry-in of the winter flood established before the run window;",
                    "held at %d mm every %d d until the WTD-derived drain."),
              IRRIGATION_MM, INTERVAL_D)
    } else if (!is.null(sp$drain)) {
      sprintf("Maintains the flooded period begun %s; ends %s, the WTD-derived drain date.",
              start, end)
    } else {
      sprintf("Maintains the flooded period begun %s to the end of the run window.", start)
    }
    new[[length(new) + 1L]] <- e
    n <- n + 1L
    d <- d + INTERVAL_D
  }
  cat(sprintf("  %s -> %s  (%3d d)  %2d events\n", start, end, as.integer(end - start), n))
}

all_ev <- c(other, new)
ord <- order(vapply(all_ev, function(e) e$date, character(1)),
             vapply(all_ev, function(e) e$event_type, character(1)))
doc$events <- all_ev[ord]
write(jsonlite::toJSON(doc, auto_unbox = TRUE, pretty = 1, digits = NA), path)

cat(sprintf("\n  irrigation: %d -> %d   total: %d -> %d\n",
            length(irr), length(new), length(ev), length(doc$events)))
